src/listener/to_listener.rs

```diff
@@ -0,0 +1,405 @@
+/// # Other implementations
+/// See below for additional provided implementations of ToListener.
+
+pub trait ToListener<State: Send + Sync + 'static> {
+    type Listener: Listener<State>;
+    /// Transform self into a
+    /// [`Listener`](crate::listener::Listener). Unless self is
+    /// already bound/connected to the underlying io, converting to a
+    /// listener does not initiate a connection. An Err return
+    /// indicates an unsuccessful conversion to a listener, not an
+    /// unsuccessful bind attempt.
+    fn to_listener(self) -> io::Result<Self::Listener>;
+}
+
+impl<State: Send + Sync + 'static> ToListener<State> for Url {
+    type Listener = ParsedListener;
+
+    fn to_listener(self) -> io::Result<Self::Listener> {
+        match self.scheme() {
+            "unix" | "file" | "http+unix" => {
+                #[cfg(unix)]
+                {
+                    let path = std::path::PathBuf::from(format!(
+                        "{}{}",
+                        self.domain().unwrap_or_default(),
+                        self.path()
+                    ));
+
+                    Ok(ParsedListener::Unix(UnixListener::from_path(path)))
+                }
+
+                #[cfg(not(unix))]
+                {
+                    Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        "Unix sockets not supported on this platform",
+                    ))
+                }
+            }
+
+            "tcp" | "http" => Ok(ParsedListener::Tcp(TcpListener::from_addrs(
+                self.socket_addrs(|| Some(80))?,
+            ))),
+
+            "tls" | "ssl" | "https" => Err(io::Error::new(
+                io::ErrorKind::Other,
+                "parsing TLS listeners not supported yet",
+            )),
+
+            _ => Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "unrecognized url scheme",
+            )),
+        }
+    }
+}
+
+impl<State: Send + Sync + 'static> ToListener<State> for String {
+    type Listener = ParsedListener;
+    fn to_listener(self) -> io::Result<Self::Listener> {
+        ToListener::<State>::to_listener(self.as_str())
+    }
+}
+
+impl<State: Send + Sync + 'static> ToListener<State> for &str {
+    type Listener = ParsedListener;
+
+    fn to_listener(self) -> io::Result<Self::Listener> {
+        self.to_socket_addrs()
+            .and_then(|socket_addrs| {
+                Ok(ParsedListener::Tcp(TcpListener::from_addrs(
+                    socket_addrs.collect(),
+                )))
+            })
+            .or_else(|_| {
+                Url::parse(self)
+                    .map_err(|_| {
+                        io::Error::new(
+                            io::ErrorKind::InvalidInput,
+                            format!("unable to parse listener from `{}`", self),
+                        )
+                    })
+                    .and_then(ToListener::<State>::to_listener)
+            })
+    }
```

Comment:
This is quite hard to follow. Using `match` seems preferable.

Reply:
Totally agreed. I dislike this code a lot and went through a bunch of iterations of it. I'll see what I can do to improve it

Reply:
I went to suggest that but I wasn't sure since the match doesn't end up being much shorter... is probably still easier to read though. Maybe this could be clarified even more:

```rust
if let Some(socket_addrs) = self.to_socket_addrs()
    Ok(ParsedListener::Tcp(TcpListener::from_addrs(
        socket_addrs.collect(),
    )))
} else {
    Url::parse(self)
        .map_err(|_| {
            io::Error::new(
                io::ErrorKind::InvalidInput,
                format!("unable to parse listener from `{}`", self),
             )
         })
         .and_then(ToListener::<State>::to_listener)
}
```